Replace `RollingSessionWindow` in approval-voting with `RuntimeInfo`

tdimitrov opened the pull request:
closes https://github.com/paritytech/polkadot/issues/6144

This is the 2nd change for https://github.com/paritytech/polkadot/issues/6812. It's similar to https://github.com/paritytech/polkadot/pull/6968 but addresses approval-voting.

Please refer to the description in https://github.com/paritytech/polkadot/pull/6968 for some context.

This PR introduces `SessionInfoProvider` which wraps `RuntimeInfo` and some additional fields for book keeping. Unlike `dispute-coordinator`, `approval-voting` doesn't wait for the first leaf to initialize its session information and directly starts its main loop. For this reason `RollingSessionWindow` was wrapped in `Option`. It is `None` on startup and gets initialized on first leaf update. I can't see a reason for this to be honest but it probably handles some edge case so I kept this behaviour. **If this is not the case - please leave a comment**. Removing this might make the overall code a bit nicer.

`RollingSessionWindow` was part of `State. However `RuntimeInfo` requires `&mut self` for each operation which created some complications so I decided to separate them. Furthermore this allowed me to remove some mut borrows of `State` throughout the code.

`RuntimeInfo` is async which required to convert some of the functions in `approval-voting` to `async` ones but this change is small and self-explanatory.

Another change worth mentioning is that `session_info` (renamed to `get_session_info`) and `cache_session_info_for_head` are now free standing functions, but more or less their behaviour is the same. `cache_session_info_for_head` pre-caches `SessionInfo` in a fashion similar to the one used in `dispute-coordinator`.

tdimitrov commented on node/core/approval-voting/src/lib.rs:
```diff
@@ -779,16 +763,15 @@ where
 	};
 
+	// `None` on start-up. Gets initialized/updated on leaf update
+	let mut session_info_provider: Option<SessionInfoProvider> = None;
```
We can drop the `Option` if it's safe to block the subsystem until we get the first `ActiveLeavesUpdate`.

eskimor replied:
No need to block. `RuntimeInfo` does not need an active leaf for [creation](https://github.com/paritytech/polkadot/blob/a7b74ccbfd520146765e20cca9641f7e4ae8e213/node/subsystem-util/src/runtime/mod.rs#L105).

ordian commented:
While it's ok for dispute-coordinator to only store the past `dispute_period` sessions, for approval-voting, we want to go as deep as the last finalized block. See also: https://github.com/paritytech/polkadot/issues/6165

eskimor commented on node/core/approval-voting/src/import.rs:
```diff
@@ -882,13 +918,30 @@ pub(crate) mod tests {
+				runtime_info: RuntimeInfo::new_with_cache(
+					RuntimeInfoConfig {
+						keystore: None,
+						session_cache_lru_size: NonZeroUsize::new(DISPUTE_WINDOW.get() as usize)
```
`SessionWindowSize` should have a conversion function to `NonZeroUsize` - or alternatively even better be an `NonZeroUsize` newtype wrapper - if that's possible.

tdimitrov replied:
I've added a conversion function. I've tried wrapping `NonZeroUsize` instead of `SessionIndex` but I hit a wall with
https://github.com/paritytech/polkadot/blob/233478e1987886c58e8f9030a6fa221d43035f82/node/primitives/src/lib.rs#L141
For `NonZeroUsize` it's an unsafe function so this won't work without an `expect()`. Maybe we can do something smarter but better handle it separately.

Should I open an issue for this?

eskimor replied:
If you hit a wall it is probably not worth pursuing right now. Thanks for checking!

eskimor commented on node/core/approval-voting/src/lib.rs:
```diff
@@ -2612,7 +2679,14 @@ async fn issue_approval<Context>(
 	issue_approval_span.add_int_tag("candidate_index", candidate_index as i64);
 
-	let session_info = match state.session_info(block_entry.session()) {
+	let session_info = match get_session_info(
```
`get_sesion_info` could be a member function of `SessionInfoProvider` once we got rid of the wrapping `Option`.

eskimor commented on node/core/approval-voting/src/lib.rs:
```diff
@@ -2760,3 +2837,158 @@ fn issue_local_invalid_statement<Sender>(
+	match session_info.take() {
+		None => {
+			let mut runtime_info = RuntimeInfo::new_with_config(RuntimeInfoConfig {
+				keystore: None,
```
Have not checked, but is there really no use for [`ValidatorInfo`](https://github.com/paritytech/polkadot/blob/a7b74ccbfd520146765e20cca9641f7e4ae8e213/node/subsystem-util/src/runtime/mod.rs#L86) in this subsystem? Maybe some code can be simplified by using this?

tdimitrov replied:
I don't see any usages of `ValidatorInfo` but what simplification you have got in mind?

tdimitrov commented:
> While it's ok for dispute-coordinator to only store the past `dispute_period` sessions, for approval-voting, we want to go as deep as the last finalized block. See also: #6165

I see what you mean but with `RuntimeInfo` we are just caching, so if we don't cache something we can look it up later as long as it's not pruned. And if `approval-voting` works on unfinalized data - this should not be the case, at least generally speaking. So I guess the biggest problem is the case when we work on a candidate, its corresponding `SessionInfo` is not cached for some reason, it gets finalized meanwhile (and pruned).

Am I thinking in the right direction?

sandreim commented:
The runtime keeps only 6 sessions in state. In the case of a long finality stall we would need to query session infos from older blocks. See https://github.com/paritytech/polkadot/pull/7127

tdimitrov commented:
> The runtime keeps only 6 sessions in state. In the case of a long finality stall we would need to query session infos from older blocks. See #7127

You mean the cache in runtime api subsystem or the actual runtime?

ordian commented on node/core/approval-voting/src/lib.rs:
```diff
@@ -638,69 +637,54 @@ impl CurrentlyCheckingSet {
+
+impl SessionInfoProvider {
+	fn earliest_session(&self) -> SessionIndex {
+		self.highest_session_seen.saturating_sub(DISPUTE_WINDOW.get() - 1)
```
Again, I'm not sure why dispute window from the highest seen session is used to bail with `BlockFromAncientSession` for approval-voting. We should allow importing and producing messages related to down to the last finalized block. Or do you set `highest_session_seen` depending on each `head`?

tdimitrov replied:
I don't set `highest_session_seen` for each head. This is wrong - earliest session should be the one of the last finalized block, as you already mentioned.

tdimitrov replied:
This is also fixed. `earliest_session` is removed and the last finalized block is used instead.

ordian commented on node/core/approval-voting/src/import.rs:
```diff
@@ -360,25 +361,7 @@ pub(crate) async fn handle_new_head<Context, B: Backend>(
-		},
-		Ok(_) => {},
-	}
+	cache_session_info_for_head(ctx.sender(), head, session_info_provider).await;
```
if I'm reading this correctly, this will update the highest_seen_session and lead to bailing on ancient session error in case finality stalls for 6+ sessions?

tdimitrov replied:
Yes, but this is just caching. If the session is actually needed - it will be queried with a relay parent and should be obtainable.

tdimitrov replied:
I've removed the caching altogether.

ordian reviewed (commented):
Maybe I'm misreading the code, but I think we should make sure #6165 is addressed properly here.

tdimitrov commented:
@ordian just to be sure we are on the same page:

1. `earliset_session()` implementation is bad, because it uses dispute window, instead of the last finalized block's session.
2. Caching is not good, because it goes six blocks back. In case of stalled finality it will miss sessions (potentially a lot of sessions on testnets).
3. This is my opinion and I want to know if you agree - `get_session_info()` usage is fine because it uses block's hash (or candidate's parent hash) to query SessionInfo. Even if this session is not cached we will be able to fetch it from the runtime.

Do you agree (especially with 3)?

ordian commented:
> @ordian just to be sure we are on the same page:
> 
>     1. `earliset_session()` implementation is bad, because it uses dispute window, instead of the last finalized block's session.
> 
>     2. Caching is not good, because it goes six blocks back. In case of stalled finality it will miss sessions (potentially a lot of sessions on testnets).
> 
>     3. This is my opinion and I want to know if you agree - `get_session_info()` usage is fine because it uses block's hash (or candidate's parent hash) to query SessionInfo. Even if this session is not cached we will be able to fetch it from the runtime.
> 
> 
> Do you agree (especially with 3)?

In principle yes, I agree. But I'm not sure exactly what is the purpose of `RuntimeInfo` in this case. If we rely on `runtime-api` caching to be enough for all unfinalized blocks, what does this abstraction bring? In other words, why not simply query session info with the appropriate relay parent?

ordian commented:
> > @ordian just to be sure we are on the same page:
> > ```
> > 1. `earliset_session()` implementation is bad, because it uses dispute window, instead of the last finalized block's session.
> > 
> > 2. Caching is not good, because it goes six blocks back. In case of stalled finality it will miss sessions (potentially a lot of sessions on testnets).
> > 
> > 3. This is my opinion and I want to know if you agree - `get_session_info()` usage is fine because it uses block's hash (or candidate's parent hash) to query SessionInfo. Even if this session is not cached we will be able to fetch it from the runtime.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Do you agree (especially with 3)?
> 
> In principle yes, I agree. But I'm not sure exactly what is the purpose of `RuntimeInfo` in this case. If we rely on `runtime-api` caching to be enough for all unfinalized blocks, what does this abstraction bring? In other words, why not simply query session info with the appropriate relay parent from the runtime-api subsystem?

tdimitrov commented:
> In other words, why not simply query session info with the appropriate relay parent from the runtime-api subsystem?

I am thinking the same after the feedback I received. Caching doesn't make much sense, because the cache size is fixed while the sessions we care about can grow almost indefinitely (despite it is a strong word). So in nutshell - we can't cache what we need.

I created the abstraction as a wrapper because of the caching - to hold together `gaps_in_cache` and `highest_session_seen` together with the `RuntimeInfo` instance. It makes no sense without caching.

tdimitrov commented:
> But I'm not sure exactly what is the purpose of `RuntimeInfo` in this case.

That was supposed to be my next point. Considering the feedback I've received - I think the same. We can't cache the sessions in a sensible way so using `RuntimeInfo` (which we picked for `dispute-coordinator` because of the controllable cache size if I've understood correctly) doesn't make much sense.

If there are no other concerns I'll use the runtime api subsystem directly.

sandreim commented:
Maybe its enough to just bump `session_info LruCache` size to something that brings same benefits as rolling session window does, but without the complexity.

sandreim commented:
And also `session_index_for_child`. If finality stalls for long, block production would slow down as well as parachains, so we don't care anymore about being efficient.

tdimitrov commented:
> Maybe its enough to just bump `session_info LruCache` size to something that brings same benefits as rolling session window does, but without the complexity.

I thought about this but If I remember correctly the count of the saved sessions is dynamic in `RollingSessionWindow`. If we do the same here, we'll save a lot of unnecessary data which furthermore won't be used on production.

eskimor commented:
> > But I'm not sure exactly what is the purpose of `RuntimeInfo` in this case.
> 
> That was supposed to be my next point. Considering the feedback I've received - I think the same. We can't cache the sessions in a sensible way so using `RuntimeInfo` (which we picked for `dispute-coordinator` because of the controllable cache size if I've understood correctly) doesn't make much sense.
> 
> If there are no other concerns I'll use the runtime api subsystem directly.

The `RuntimeAPI` has two purposes:

1. A more ergonomic API.
2. Additional local caching.

I envisioned it to be the canonical way of accessing session related information. I don't see any advantage in using the runtime api directly.

eskimor commented:
> > Maybe its enough to just bump `session_info LruCache` size to something that brings same benefits as rolling session window does, but without the complexity.
> 
> I thought about this but If I remember correctly the count of the saved sessions is dynamic in `RollingSessionWindow`. If we do the same here, we'll save a lot of unnecessary data which furthermore won't be used on production.

I think we should just assume that 6 sessions are enough. Especially for approval voting: Why would we ever still be busy with approval voting on a block that is 6 sessions old? Even on test networks this is hard to imagine to ever happen.

ordian commented:
> > > Maybe its enough to just bump `session_info LruCache` size to something that brings same benefits as rolling session window does, but without the complexity.
> > 
> > 
> > I thought about this but If I remember correctly the count of the saved sessions is dynamic in `RollingSessionWindow`. If we do the same here, we'll save a lot of unnecessary data which furthermore won't be used on production.
> 
> I think we should just assume that 6 sessions are enough. Especially for approval voting: Why would we ever still be busy with approval voting on a block that is 6 sessions old? Even on test networks this is hard to imagine to ever happen.

We had this problem on Westend a few days ago :P https://github.com/paritytech/polkadot/pull/7127 did help recover it.

tdimitrov commented:
> A more ergonomic API.

I haven't got a strong opinion on this. We only fetch sessions by index in approval voting so it's not a big deal to use the runtime-api directly. At the same time I don't mind using `RuntimeInfo` as it will be consistent with `dispute-coordinator` and `dispute-distribution`. 

> Additional local caching.

The way I see it now - we work only on unfinalized sessions and fetch them via relay parent hash. Caching is not helping as in `dispute-coordinator`.


> We had this problem on Westend a few days ago :P https://github.com/paritytech/polkadot/pull/7127 did help recover it.

I feel I'm not understanding something because you keep on raising this topic :) This is the default way we fetch sessions in `approval-voting` (with parent hash instead of active leaf). Why it won't work here?

tdimitrov commented:
> Requesting the whole SessionInfo over and over again is certainly not super cheap. 

So this is what you mean by local cache. Not sure if it will be a performance bottle neck but it's true. But is it worth pre-caching the sessions?

ordian commented:
> Then we are golden, as I believe the relevant relay parent is used and not the latest. I think we are good. In approval voting we also don't care about dead forks that fell behind finality and could have pruned state.

Not quite. The problem is with `earliest_session` [here](https://github.com/paritytech/polkadot/pull/7123#discussion_r1176488333) which will make it exit early even before requesting the session info. 

I suggest to remove this method altogether and instead keep track of latest finalized block number and look at that (bail if < finalized_number).

> This is the default way we fetch sessions in approval-voting (with parent hash instead of active leaf). Why it won't work here?

Let's say we're major syncing and our parachain db is empty. We received a finalized notification of block at height 1000 and active leaf at 2000. Before #7127, we'd request all needed session infos using block 2000 as parent. Which would only give us latest 6 sessions.  After #7127, we'd requrest session info using block 1001, 1002, ... 2000, which fixes the problem.
With the current state of your PR, we'd cache 6 sessions using 2000 and then bail due to `earliest_session` mentioned above.

tdimitrov commented:
Okay, what you wrote above is sort of correct. We already [agreed](https://github.com/paritytech/polkadot/pull/7123#issuecomment-1523117304) it is wrong. `earliest_session` is getting fixed/removed no matter what.

But the scenario you described is only valid for the session caching. The actual session fetching in the rest of the code (in this PR) is made via parent hash so it should be fine.

I think I understand the concerns of both of you - let me rework the PR and we'll continue discussing the final solution.

tdimitrov commented:
I've removed the check for earliest session and check if the block is finalized instead.

Also I removed the pre-caching. I think it won't help much because of the reasons @ordian and @sandreim - in case of a finality lag we won't be able to fill in the cache. During normal operation - the cache will be populated the first time when a session is requested.

The last two commits are work in progress and need polishing and fixing the tests. Please ignore (for now) things like commented out code blocks in the tests, misleading variable names and broken tests. I want to get some feedback first because adjusting the tests will be time  consuming and I want to do it only once :)

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2747951

eskimor commented:
> I've removed the check for earliest session and check if the block is finalized instead.
> 
> Also I removed the pre-caching. I think it won't help much because of the reasons @ordian and @sandreim - in case of a finality lag we won't be able to fill in the cache. During normal operation - the cache will be populated the first time when a session is requested. Any thoughts on this @eskimor?
> 
> The last two commits are work in progress and need polishing and fixing the tests. Please ignore (for now) things like commented out code blocks in the tests, misleading variable names and broken tests. I want to get some feedback first because adjusting the tests will be time consuming and I want to do it only once :)

Sounds sensible.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2765898

ordian commented on node/core/approval-voting/src/import.rs:
```diff
@@ -160,19 +159,17 @@ async fn imported_block_info<Context>(
-			.map_or(true, |s| session_index < s.earliest_session())
-		{
+		// We can't determine if the block is finalized or not - try processing it
+		if last_finalized_height.map_or(false, |finalized| block_header.number < finalized) {
```
:+1:

ordian reviewed (commented):
Looks nice except for the log levels which need double checking.

eskimor commented on node/core/approval-voting/src/import.rs:
```diff
@@ -461,11 +442,25 @@ pub(crate) async fn handle_new_head<Context, B: Backend>(
+		{
+			Some(session_info) => session_info,
+			None => {
+				gum::error!(
```
again `get_session_info` is already logging. If we want more precise logs, e.g. saying that it was for a new head: That is fine as well, but then we should reallly get rid of `get_session_info` as it does nothing useful then.

tdimitrov replied:
Fixed for all `get_session_info()` calls. Only the log line inside the function is kept.

eskimor reviewed (approved):
Looks good! Actual removal of RollingSessionWindow and db migration will be a separate PR?

tdimitrov commented on node/core/approval-voting/src/import.rs:
```diff
@@ -212,15 +209,19 @@ async fn imported_block_info<Context>(
+		match get_session_info(env.runtime_info, ctx.sender(), block_hash, session_index).await {
+			Some(session_info) => session_info,
+			None => {
+				gum::error!(
```
I agree with both. I missed the double-logging and logging session info errors at ERROR level will be too noisy. Fixed both.

tdimitrov commented:
I've fixed all outstanding issues. I'll do a burnin on versi and this can be merged.

> Actual removal of RollingSessionWindow and db migration will be a separate PR?

My plan was to remove `RollingSessionWindow` as a separate PR but I forgot about the storage migration. I'll create a new PR on top of this one (to keep the diffs clean) and will merge them together after the burnin.

ordian reviewed (approved):
Good job!

eskimor commented on node/core/approval-voting/src/import.rs:
```diff
@@ -212,15 +212,7 @@ async fn imported_block_info<Context>(
 	let session_info =
 		match get_session_info(env.runtime_info, ctx.sender(), block_hash, session_index).await {
 			Some(session_info) => session_info,
-			None => {
```
Now you can do a simple : ok_or(...)?;

tdimitrov replied:
Fixed.

eskimor commented on node/core/approval-voting/src/import.rs:
```diff
@@ -451,15 +443,7 @@ pub(crate) async fn handle_new_head<Context, B: Backend>(
-					?head,
-					"Can't get session info for the new head"
-				);
-				return Ok(Vec::new())
```
unwrap_or would be nicer.

eskimor replied:
Yeah you are right, my mistake.
